symbolic-base/src/ZkFold/Symbolic/Compiler.hs

```diff
@@ -44,85 +47,86 @@ import           ZkFold.Symbolic.MonadCircuit               (MonadCircuit (..))
+  , SymbolicInput s, Context s ~ c, Symbolic c
+  ) => c (Layout s) -> f -> c (Layout f)
+solderWith sLayout f = fromCircuit2F (pieces f input) b $ \r (Par1 i) -> do
+  constraint (\x -> one - x i)
+  return r
   where
     Bool b = isValid input
-    input = restore @(Support f) $ const idCircuit
+    input = restore (const sLayout)
+
+compileCircuit ::
+  ( SymbolicData y, Context y ~ c, Support y ~ Proxy c, Layout y ~ l
+  , c ~ ArithmeticCircuit a p i
+  ) => c l -> y
+compileCircuit = restore . const . optimize
+
+solder ::
+  ( SymbolicData f, Context f ~ c, Support f ~ s
+  , SymbolicInput s, Context s ~ c, Layout s ~ l
+  , c ~ ArithmeticCircuit a p l, Arithmetic a, Binary a, Binary (Rep p)
+  ) => f -> c (Layout f)
+solder = solderWith idCircuit
 
 -- | Compiles function `f` into an arithmetic circuit with all outputs equal to 1.
 compileForceOne ::
-    forall a c p f s l y .
-    ( c ~ ArithmeticCircuit a p l
-    , Arithmetic a
-    , Binary a
-    , SymbolicData f
-    , Context f ~ c
-    , Support f ~ s
-    , SymbolicInput s
-    , Context s ~ c
-    , Layout s ~ l
-    , Binary (Rep p)
-    , Binary (Rep l)
-    , Ord (Rep l)
-    , SymbolicData y
-    , Context y ~ c
-    , Support y ~ Proxy c
-    , Layout f ~ Layout y
-    , Traversable (Layout y)
-    ) => f -> y
-compileForceOne = restore . const . optimize . forceOne . solder @a
+  forall a c p f s l y .
+  ( c ~ ArithmeticCircuit a p l, Arithmetic a, Binary a
+  , Binary (Rep p), Binary (Rep l), Ord (Rep l)
+  , SymbolicData f, Context f ~ c, Support f ~ s
+  , SymbolicInput s, Context s ~ c, Layout s ~ l
+  , SymbolicData y, Context y ~ c, Support y ~ Proxy c
+  , Layout f ~ Layout y, Traversable (Layout y)
+  ) => f -> y
+compileForceOne = compileCircuit . forceOne . solder
 
 -- | Compiles function `f` into an arithmetic circuit.
 compile ::
-    forall a c p f s l y .
-    ( c ~ ArithmeticCircuit a p l
-    , SymbolicData f
-    , Context f ~ c
-    , Support f ~ s
-    , SymbolicInput s
-    , Context s ~ c
-    , Layout s ~ l
-    , SymbolicData y
-    , Context y ~ c
-    , Support y ~ Proxy c
-    , Layout f ~ Layout y
-    , Symbolic c
-    ) => f -> y
-compile = restore . const . optimize . solder @a
+  forall a p c f s l y .
+  ( c ~ ArithmeticCircuit a p l, Arithmetic a, Binary a, Binary (Rep p)
+  , SymbolicData f, Context f ~ c, Support f ~ s
+  , SymbolicInput s, Context s ~ c, Layout s ~ l
+  , SymbolicData y, Context y ~ c, Support y ~ Proxy c
+  , Layout f ~ Layout y
+  ) => f -> y
+compile = compileCircuit . solder
+
+compileWithPayload ::
+  forall a p c f s l y .
+  ( c ~ ArithmeticCircuit a l p, Arithmetic a, Binary a
+  , Binary (Rep p), Ord (Rep p), NFData (Rep p)
+  , Traversable l
+  , SymbolicData f, Context f ~ c, Support f ~ s
+  , SymbolicInput s, Context s ~ c, Layout s ~ l
+  , SymbolicData y, Context y ~ c, Support y ~ Proxy c
+  , Layout f ~ Layout y
+  ) => f -> y
+compileWithPayload = compileCircuit . solderWith payloadCircuit
```

Comment:
Proposal: let's consider having a single `compile` function with an additional "options" argument. Would it help avoid the mistake of selecting a "wrong" compile function?

Reply:
Well, we can leave only the `compileWith f` function which would look like `compileCircuit . f`, where `f` can be a `solderWith c` postcomposed with some desired operations

Reply:
Left only `solderWith`, `finalRestore`, `compile` and `compileIO` in the public API